APE22.rst

```diff
@@ -0,0 +1,440 @@
+     considered private.
+   
+   - A module can expose an ``__all__`` symbol at the module level that provides
+     a list of names that are considered part of the interface. This overrides
+     all other rules above, allowing imported symbols or symbols whose names
+     begin with an underscore to be included in the interface.
+
+This is consistent with PEP 8 and SciPy, but it does not mention the
+documentation. The Python typing guide reinforces the idea that the ``__all__``
+attribute is the authoritative source of what is public vs internal.
+Importantly, the ``__all__`` attribute is a module-level attribute, so it only
+applies to the module in which it is defined. We will refer to this as "locally"
+public. This means that a module can define an ``__all__`` attribute but if the
+module itself is not public, then anything in the ``__all__`` attribute cannot
+be publicly accessed from outside the module.
+
+
+Solution
+--------
+
+We propose the following:
+
+1. That Astropy adopts the PEP 8 rules on public versus internal interfaces
+2. That Astropy disambiguate these rules similarly to SciPy, by adopting the
+   subsequent rules for its API:
+3. That Astropy ensures its documentation is consistent with its code, the
+   latter being the authoritative source.
+4. That Astropy strongly recommends these rules for coordinated packages, and
+   encourage affiliated packages to follow these rules as well.
+
+**Rules for Public Interfaces:**
+
+1. A symbol is public if all containing namespaces are public.
+   A symbol is internal if any containing namespace is internal.
+
+2. All modules must have an ``__all__`` attribute, even if it is empty. The
+   ``__all__`` attribute specifies the public and internal interface of the
+   module in which it is defined. Anything in ``__all__`` is *locally* public,
+   including underscore-prefixed symbols. Anything not in ``__all__`` is
+   internal in that module. The exception to this rule are modules that cannot
+   have an ``__all__`` attribute, for example implicit namespace packages, which
+   lack an ``__init__.py``. In these cases, the public interface is defined by
+   the sub-rules:
+
+   - A symbol is *locally* public if it does not start with an underscore, and
+     internal otherwise, unless it is a dunder symbol (``__<...>__``).
+   - Dunder symbols (``__<...>__``) have their own public vs internal rules,
+     which are not determined in this APE.
+
+3. Public symbols must not be prefixed with an underscore.
+
+   - Internal symbols need not be prefixed with an underscore, but it is often
+     recommended. An example of when it is not needed are symbols defined in
+     internal modules, making the prefix redundant.
+
+4. Public symbols must be documented.
+
+   - If an internal symbol is documented (which is not recommended), it must be
+     obviously, explicitly, (and preferably repeatedly) noted as internal.
```

Comment:
"obvious" may sometimes give off a condescending vibe to the reader, let's avoid that
```suggestion
     unambiguously, explicitly, (and preferably repeatedly) noted as internal.
```